support for localhost/127.0.0.1 proxy bypass

elerch opened the pull request:
This PR addresses #1565 and #2130. It is currently using hardcoded values for localhost (localhost and 127.0.0.1) for bypass. I started wiring this into the settings system but was getting seg faults so I'm clearly doing something wrong there. I'll create a new PR for customization of the bypass values through the settings once I get that worked out.

This was originally worked off rev 7f74d893635d275f9450bf5f12c454a9c27672bd.

elerch commented:
I've added two commits for the settings. The issue I had seems to be build related - performing a clean build removed the segfault.

The first settings commit adds settings, seeding the list with "localhost" and "127.0.0.1". The second reverts that change but instead detects an empty list and uses "localhost" and "127.0.0.1". This allows you to use a command line flag like --proxy-bypass none, which is technically a host but shouldn't match anything and will therefore force all traffic through the proxy. However, if you completely omit --proxy-bypass, you'll get behavior you probably expect (use the proxy except for localhost/127.0.0.1).

I considered the second change slightly controversial so I've not squashed these commits.

ashkulz commented on src/lib/multipageloader.cc:
```diff
@@ -107,6 +107,20 @@ QNetworkReply * MyNetworkAccessManager::createRequest(Operation op, const QNetwo
 	return QNetworkAccessManager::createRequest(op, r3, outgoingData);
 }
 
+MyNetworkProxyFactory::MyNetworkProxyFactory (QNetworkProxy proxy, QList<QString> bph, ResourceObject & res):
```
Why do you need the `ResourceObject` here? It isn't used, so should ideally be removed.

elerch replied:
Just an oversight - I was originally using it, then removed it but missed this line. Thanks!

ashkulz commented on src/lib/multipageloader.cc:
```diff
@@ -107,6 +107,20 @@ QNetworkReply * MyNetworkAccessManager::createRequest(Operation op, const QNetwo
+QList<QNetworkProxy> MyNetworkProxyFactory::queryProxy (const QNetworkProxyQuery & query) {
+	QString host = query.url().host();
+	foreach (const QString & bypassHost, bypassHosts) {
+		if (host == bypassHost) {
```
Should ideally use `host.compare(bypassHost, Qt::CaseInsensitive)` as then `google.com` and `GOOGLE.COM` will match -- DNS is normally case-insensitive.

ashkulz commented:
Also, I don't think the approach taken is ok (bypass proxy for localhost by default). If someone wants that, it is better to opt in via `--proxy-bypass localhost` instead of having to opt-out via `--proxy-bypass none` (which will bypass the proxy for a hostname of `none`). Also, I think using command-line option of `--bypass-proxy-for` would be better than `--proxy-bypass`.

Thanks for the contribution, though!

ashkulz commented:
Also, please squash the commits and rebase it against master and add an entry in both `AUTHORS` and `CHANGELOG.md` mentioning the issue: put 1565 in the changelog but the commit message should refer to both issues, see e02ff7efaf7b617986fc11c08a693e203c64a465 as an example.

ashkulz commented:
@elerch: any update?

elerch commented:
Sorry - got pulled into some fires at the office. Things are dying down so I just got back to this and should be able to make the changes tomorrow.

elerch commented:
I squashed the commit, made the changes you suggested and rebased on the latest commits. The changes were retested after a build but with Qt upgraded it's probably worth a clean build, so I'm doing that now. This is probably ready to go but I'll do final testing once the build is complete. Let me know if there's any other changes you think I need to make. Thanks!

elerch commented:
FYI - Finished re-testing on a clean build and all is well, so this is ready as far as I'm concerned. Let me know if you want any other changes.

ashkulz commented:
Looks good, once you make the above changes I'll merge it ASAP.

ashkulz commented:
Also, would prefer if you move up your name so that it is in the list of `AUTHORS` who have an email address.

elerch commented:
Your comments all made sense. I made the corresponding changes and retested.

ashkulz commented:
Sorry, a few more nitpicks ... thought you'd remove the unused variable. Also, I think that the commit message should be `support bypassing the proxy for specific hosts` (i.e. no repeated for)

elerch commented:
no worries - removed the unused variable and changed the message

ashkulz commented:
Thank you for the contribution, and appreciate your patience during the review process.

monken commented:
It would be great if you could honor the `no_proxy` environmental variable which exists exactly for this purpose and is widely used.

ashkulz commented:
I'd appreciate a PR which implements this.

ashkulz commented:
[`0.12.3`](http://wkhtmltopdf.org/downloads.html#stable) has been released, which should contain the fix for this issue. Please report back if it is not solved with the above version.
